app/src/main/java/ch/epfl/cs311/wanderwave/ui/components/utils/BeaconPlacer.kt

```diff
@@ -0,0 +1,98 @@
+package ch.epfl.cs311.wanderwave.ui.components.utils
+
+import ch.epfl.cs311.wanderwave.model.data.Beacon
+import ch.epfl.cs311.wanderwave.model.data.Location
+import com.google.android.gms.maps.model.LatLng
+import kotlin.math.*
+import kotlin.random.Random
+
+private const val EARTH_RADIUS_M = 6371000.0
+private const val BEACON_RADIUS = 1000.0
+private const val BEACON_COUNT = 20
+private const val NUMBER_ITERATION = 5
+
+fun placeBeaconsRandomly(beacons: List<Beacon>, location: LatLng): List<Beacon> {
+  var finalBeacons = mutableListOf<Beacon>()
+  var nearbyBeacons = findNearbyBeacons(location, beacons, BEACON_RADIUS)
+  if (nearbyBeacons.size < BEACON_COUNT) {
+    var currentMaxDistance = 0.0
+    repeat(NUMBER_ITERATION) {
+      val newBeacons = mutableListOf<Beacon>()
+      repeat(BEACON_COUNT - nearbyBeacons.size) { findRandomBeacon(location, newBeacons, it) }
+      var newMaxDistance = computeDistanceBetweenBeacons(newBeacons, beacons)
+      if (newMaxDistance > currentMaxDistance || finalBeacons.isEmpty()) {
+        currentMaxDistance = newMaxDistance
+        finalBeacons = newBeacons
+      }
+    }
+    // TODO: add the beacons to the repo or firebase or whatever
+    // These are the beacons that are to be added : finalBeacons
+  }
+
+  return finalBeacons
+}
+
+fun computeDistanceBetweenBeacons(newBeacons: MutableList<Beacon>, beacons: List<Beacon>): Double {
+  var distance = 0.0
+  newBeacons.forEach { beacon ->
+    beacons.forEach { existingBeacon ->
+      distance += haversine(beacon.location.toLatLng(), existingBeacon.location.toLatLng())
+    }
+  }
+  return distance
+}
+
+fun findRandomBeacon(location: LatLng, newBeacons: MutableList<Beacon>, it: Int) {
+  val randomLocation = randomLatLongFromPosition(location, BEACON_RADIUS)
+  val newBeacon =
+      Beacon(
+          id = "beacon${newBeacons.size + it}", // TODO: modify the ID!!!!!!!!
```

Comment:
Put an empty id and add it to firebase, I can create you a function that will add the beacon to firebase and return the new ID. Tell me if you need it, I'll push it on your branch if you need :)

Reply:
Yes please, it will be much appreciated